session.go

```diff
@@ -511,11 +487,26 @@ type fixIn struct {
 	receiveTime time.Time
 }
 
+type sendRequest struct {
+	msg Message
+	err chan error
+}
+
 func (s *session) run(msgIn chan fixIn, msgOut chan []byte, quit chan bool) {
+	s.messageIn = msgIn
 	s.messageOut = msgOut
+	s.toSend = make(chan sendRequest)
+	s.resendIn = make(chan Message, 1)
+
+	type fromCallback struct {
```

Comment:
the size of golang structs I believe are just 8 bytes in addition to the types they enclose. So every time you'd call `run`, you'd be allocating, on the stack, 8 bytes + size of `Message` + size of `MessageRejectError`. Which is negligible considering when and how many times `run` is expected to be called.